### ISSUE TITLE
Standard Output Issue Causing Tests to Never End

### ISSUE THREAD

**Issue by JamiePed** [shouldly/shouldly#431]
I am fully aware that this is probably the most niche/strange issue that anyone has ever raised, so bare with me. 

I'm currently working on a program which needs to run `dotnet` commands from the command line, and then parse the results into an understandable format. However if you run `dotnet test` in a `Process` with `RedirectStandardOutput = true` and two or more tests fail, then the process will never exit and the standard output will freeze after reporting the second test failure.
 
If one or less tests fail then no problems occur, likewise if you stop redirecting the standard output the problem will not occur. 

I know this has something to do with either XUnit or Shouldly because I can only replicate this problem when I use the two in conjunction. I can't replicate this problem with xunit (without Shouldly) or MSTest (with/out Shouldly) but I can with XUnit and Shouldly. So I thought I would raise it here to start with, I'm happy to raise it with XUnit if you think that is where the problem lies!

I realise that this isn't a problem a lot of people are going to have - and I've found a workaround - but I thought I would raise the issue just in case it was a sign of another bug that could cause more widespread problems in the future.

## Replication

1. Create an XUnit test project 
2. Populate a test class with at least 2 failing tests
```csharp
namespace UnitTests.XUnit.Shouldly
{
    public class ShouldlyTests
    {
        [Fact]
        public void FailingTest1()
        {
            1.ShouldBe(2);
        }

        [Fact]
        public void FailingTest2()
        {
            "Hello".ShouldBe("World");
        }

        [Fact]
        public void PassingTest1()
        {
            true.ShouldBeTrue();
        }

        [Fact]
        public void PassingTest2()
        {
           "Not Null".ShouldNotBeNull();
        }
    }
}
```
3. Create a separate console application project
4. Call dotnet test in a separate process...
```csharp
public class Program
{
    static void Main(string[] args)
    {
        var startingInfo = new ProcessStartInfo()
        {
            FileName = "cmd.exe",
            Arguments = @"/C dotnet test",

            UseShellExecute = false,
            RedirectStandardOutput = true,

            CreateNoWindow = true,

            WorkingDirectory = @"D:\VS\SandBox\UnitTestProject\UnitTests.XUnit.Shouldly",
        };

        var process = new Process()
        {
            StartInfo = startingInfo,
        };

        process.Start();
        
        while (!process.HasExited)
        {
            Console.WriteLine("Waiting For Process...");
            Thread.Sleep(1000);
        }

        Console.WriteLine(process.StandardOutput.ReadToEnd());
        Console.ReadKey();
    }
}
```
Executing that process should never end, and will continue to print "Waiting for Process...". 
You should also be able to notice dotnet running a couple of background processes in task manager. 

If you kill the process from task manager then the output displays all the way up until the end of the stack trace for the second test failure.

That is the main problem, the task will never end if you are redirecting the output and processing it after the fact.

### Temporary Fixes 

On the off chance that anyone else has encountered this problem in the wild then I shall share how I'm getting around the problem.

1) Don't redirect the output, run the command `dotnet test --logger "trx;LogFileName=ResultsFile.trx"` and then when the process ends search for and read that file.

2) Something I discovered while trying to write this issue... continue to read the output while the process is still running. E.g.

```csharp
public class Program
{
    static void Main(string[] args)
    {
        var startingInfo = new ProcessStartInfo()
        {
            FileName = "cmd.exe",
            Arguments = @"/C dotnet test",

            UseShellExecute = false,
            RedirectStandardOutput = true,

            CreateNoWindow = true,

            WorkingDirectory = @"D:\VS\SandBox\UnitTestProject\UnitTests.XUnit.Shouldly",
        };

        var process = new Process()
        {
            StartInfo = startingInfo,
        };

        process.Start();

        while (!process.HasExited)
        {
            Console.WriteLine(process.StandardOutput.ReadLine());
        }

        Console.WriteLine(process.StandardOutput.ReadToEnd());
        Console.ReadKey();
    }
}
``` 
This stops the process running continuously. And explains why this problem can not be replicated by running `dotnet test` directly on the command line.

In summary, I'm not really sure if this is a bug or where it is found if it is a bug. I've just spent the past month or so (since the release of VS2017 though I think that is a coincidence) butting against this problem, and I thought it best to share. :)

**Comment by JakeGinnivan** [comment]
Unsure the issue is either shouldly or xunit, it is more likely to do with capturing process output. It is super easy to deadlock when doing that. See http://jake.ginnivan.net/redirecting-process-output/ or https://www.hanselman.com/blog/SoManyMistakesForMeToMakeSoLittleTimecapturingStandardErrorAndStandardOutput.aspx

The msdn docs say:

> You can use asynchronous read operations to avoid these dependencies and their deadlock potential. Alternately, you can avoid the deadlock condition by creating two threads and reading the output of each stream on a separate thread

**Comment by JamiePed** [comment]
My bad! Thanks for your time!
